aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp

```diff
@@ -227,11 +229,64 @@ Tensor q_adaptive_avg_pool2d(const Tensor& input, IntArrayRef output_size) {
     return output;
   }
 }
+
+#ifdef USE_PYTORCH_QNNPACK
+Tensor qnnpack_adaptive_avg_pool2d(
+    const at::Tensor& input,
+    IntArrayRef output_size) {
+    std::array<int64_t, 2> kernel_size;
+    std::array<int64_t, 2> stride;
+    std::array<int64_t, 2> padding{0, 0};
+    bool ceil_mode{false};
+    bool count_include_pad{false};
+
+    const auto output_shape = get_output_shape(input, output_size);
+    auto output_height = output_shape[output_shape.size() - 2];
+    auto output_width = output_shape[output_shape.size() - 1];
+    auto input_height = input.sizes()[input.dim() - 2];
+    auto input_width = input.sizes()[input.dim() - 1];
+    stride[0] = input_height / output_height;
+    stride[1] = input_width / output_width;
+    // Given the constraint that input_height/width % output_height/width == 0
+    // stride and kernel size are same.
+    kernel_size[0] = stride[0];
+    kernel_size[1] = stride[1];
+
+    return at::native::qnnp_avgpool_helper::qnnpack_avg_pool2d(
+        input,
+        kernel_size,
+        stride,
+        padding,
+        ceil_mode,
+        count_include_pad,
+        c10::nullopt);
+}
+
+bool enable_qnnpack_for_ada_avgpool(
+    const at::Tensor& input,
+    IntArrayRef output_size) {
+    const auto output_shape = get_output_shape(input, output_size);
+    auto output_height = output_shape[output_shape.size() - 2];
+    auto output_width = output_shape[output_shape.size() - 1];
+    auto input_height = input.sizes()[input.dim() - 2];
+    auto input_width = input.sizes()[input.dim() - 1];
+
+    return ((input_height % output_height == 0) &&
+        (input_width % output_width) == 0);
```

Comment:
Is this true in practice?  I thought the point of adaptive average pool was that you wanted to shrink a variable-sized input into a fixed-sized output.

Reply:
So this helps us reduce this to regular avgpool. If this condition is not met we fallback to the other slow path from here, https://github.com/pytorch/pytorch/pull/36813/files#diff-c227aae55653cf48955fdf36f92378afR286.
This is because qnnpack does not have adaptive avg pool implemented. I plan to implement this in PT qnnpack.

Reply:
I understand that this is safe, but my question is how often this is true?  If this is an uncommon case, I don't think it's worth having an optimized implementation for it.

Reply:
I dont know the extent to which this is common but this is for our internal mobile vision models which I was trying to optimize. Thats what they have. And the unoptimized path is much slower.
Maybe we can remove this once I have qnnpack implementation for the adaptive avgpool but that will still be slower in degenerate cases like this.